README.md
# SSH Authorized Keys Cookbook

[![Documentation](http://img.shields.io/badge/docs-rdoc.info-blue.svg?style=flat)](http://www.rubydoc.info/github/zuazo/ssh_authorized_keys-cookbook)
[![GitHub](http://img.shields.io/badge/github-zuazo/ssh__authorized__keys--cookbook-blue.svg?style=flat)](https://github.com/zuazo/ssh_authorized_keys-cookbook)
[![License](https://img.shields.io/github/license/zuazo/ssh_authorized_keys-cookbook.svg?style=flat)](#license-and-author)

[![Cookbook Version](https://img.shields.io/cookbook/v/ssh_authorized_keys.svg?style=flat)](https://supermarket.chef.io/cookbooks/ssh_authorized_keys)
[![Dependency Status](https://img.shields.io/librariesio/github/zuazo/ssh_authorized_keys-cookbook)](https://libraries.io/github/zuazo/ssh_authorized_keys-cookbook#dependencies)
[![Code Climate maintainability](https://img.shields.io/codeclimate/maintainability/zuazo/ssh_authorized_keys-cookbook)](https://codeclimate.com/github/onddo/ssh_authorized_keys-cookbook)
[![Build Status](https://img.shields.io/travis/zuazo/ssh_authorized_keys-cookbook/1.0.0.svg?style=flat)](https://travis-ci.org/zuazo/ssh_authorized_keys-cookbook)
[![Coverage Status](https://img.shields.io/coveralls/zuazo/ssh_authorized_keys-cookbook/1.0.0.svg?style=flat)](https://coveralls.io/r/zuazo/ssh_authorized_keys-cookbook?branch=1.0.0)
[![Inline docs](https://inch-ci.org/github/zuazo/ssh_authorized_keys-cookbook.svg?branch=master&style=flat)](https://inch-ci.org/github/zuazo/ssh_authorized_keys-cookbook)

[Chef](https://www.chef.io/) cookbook to create SSH authorized keys files in user home directories.

Tries to avoid generating a corrupt file that could render your server inaccessible.

## Requirements

### Supported Platforms

This cookbook has been tested on the following platforms:

* AIX
* Amazon Linux
* Debian
* CentOS
* Fedora
* FreeBSD
* openSUSE
* Oracle
* RedHat
* SUSE
* Ubuntu

Please, [let us know](https://github.com/zuazo/ssh_authorized_keys-cookbook/issues/new?title=I%20have%20used%20it%20successfully%20on%20...) if you use it successfully on any other platform.

### Required Applications

* Chef `14` or higher.
* Ruby `2.5` or higher.

## Definitions

### ssh_authorize_key

Authorize a key for public key authentication using SSH.

**Warning:** This definition uses the [Accumulator Pattern](http://docs.chef.io/definitions.html#many-recipes-one-definition). This implies that any SSH key added using other methods (such as **keys added by hand**) will be **deleted**.

### ssh_authorize_key Properties

| Property     | Default           | Description                              |
|:-------------|:------------------|:-----------------------------------------|
| user         | `nil`             | System user **(required)**.              |
| group        | user              | System group.                            |
| home         | *calculated*      | System user home path.                   |
| key          | `nil`             | SSH public key in base64 **(required)**. |
| keytype      | `'ssh-rsa'`       | SSH key type.                            |
| comment      | *definition name* | SSH key comment.                         |
| options      | `nil`             | SSH key options as a hash.               |
| validate_key | `true`            | Enable/Disable assert_key                |

## Usage Examples

First of all, don't forget to include the `ssh_authorized_keys` cookbook as a dependency in the cookbook metadata:

```ruby
# metadata.rb
# [...]

depends 'ssh_authorized_keys'
```

You can use the `ssh_authorize_key` to authorize SSH public keys to use SSH public key authentication:

```ruby
# Bob is the admin here.

ssh_authorize_key 'bob@acme.com' do
  key 'AAAAB3NzaC1yc2EAAAADAQABAAABAQCctNyRouVDhzjiP[...]'
  user 'root'
end

ssh_authorize_key 'alice@acme.com' do
  key 'AAAAB3NzaC1yc2EAAAADAQABAAABAQCySLKbpFRGCrKU/[...]'
  user 'alice'
end
```

### Setting the SSH Key Options Field

You can set the options field as follows:

```ruby
# As the root user by default in ubuntu:
ssh_authorize_key 'bob@acme.com' do
  key 'AAAAB3NzaC1yc2EAAAADAQABAAABAQCctNyRouVDhzjiP[...]'
  user 'root'
  options(
    'no-port-forwarding' => true,
    'no-agent-forwarding' => true,
    'no-X11-forwarding' => true,
    command:
      'echo \'Please login as the user "bob" rather than the user "root".\''\
      ';echo;sleep 10'
  )
end
```

### Reading the Keys from a Data Bag

For example, from the following data bag item:

```json
{
  "id": "users",
  "bob@acme.com": {
    "key": "AAAAB3NzaC1yc2EAAAADAQABAAABAQCctNyRouVDhzjiP[...]",
    "user": "root"
  },
  "alice@acme.com": {
    "key": "AAAAB3NzaC1yc2EAAAADAQABAAABAQCySLKbpFRGCrKU/[...]",
    "user": "alice"
  }
}
```

You can read the data bag item from a recipe as follows:

```ruby
users = data_bag_item('ssh', 'users')
users.delete('id')

users.each do |name, ssh_key|
  ssh_authorize_key name do
    key ssh_key['key']
    user ssh_key['user']
  end
end
```

See [the data bags DSL documentation](http://docs.chef.io/data_bags.html#load-with-dsl-recipe) for a more detailed explanation and [the data bags knife documentation](http://docs.chef.io/data_bags.html#using-knife-title) to learn how to create a data bag.

## Attributes


These attributes are primarily intended to support the different platforms. Do not touch them unless you know what you are doing.

| Attribute                                 | Default      | Description            |
|:------------------------------------------|:-------------|:-----------------------|
| `node['ssh_authorized_keys']['keytypes']` | *calculated* | Allowed SSH key types. |

## Testing


See [TESTING.md](https://github.com/zuazo/ssh_authorized_keys-cookbook/blob/master/TESTING.md).

## ChefSpec Tests

To create ChefSpec tests for the `ssh_authorize_key` definition, you can use the [`render_file`](http://www.rubydoc.info/github/sethvargo/chefspec#render_file) matcher to check the *authorized_keys* file content:

```ruby
it 'allows bob to login as root' do
  expect(chef_run).to render_file('/root/.ssh/authorized_keys')
    .with_content(/^ssh-rsa [A-Za-z0-9+\/=]+ bob@acme\.com$/)
end
```

You can also test against the internal template:

```ruby
it 'creates ~bob/.ssh/authorized_keys file' do
  expect(chef_run).to create_template('/home/bob/.ssh/authorized_keys')
end
```

## Contributing


Please do not hesitate to [open an issue](https://github.com/zuazo/ssh_authorized_keys-cookbook/issues/new) with any questions or problems.

See [CONTRIBUTING.md](https://github.com/zuazo/ssh_authorized_keys-cookbook/blob/master/CONTRIBUTING.md).

## TODO


See [TODO.md](https://github.com/zuazo/ssh_authorized_keys-cookbook/blob/master/TODO.md).


## License and Author

|                      |                                          |
|:---------------------|:-----------------------------------------|
| **Author:**          | [Raul Rodriguez](https://github.com/raulr) (<raul@onddo.com>)
| **Author:**          | [Xabier de Zuazo](https://github.com/zuazo) (<xabier@zuazo.org>)
| **Contributor:**     | [Ong Ming Yang](https://github.com/ongmingyang)
| **Contributor:**     | [MVNW](https://github.com/MVNW)
| **Contributor:**     | [Anthony Caiafa](https://github.com/acaiafa)
| **Contributor:**     | [Benjamin M. Hughes](https://github.com/bmhughes)
| **Contributor:**     | [Corey Hemminger](https://github.com/Stromweld)
| **Copyright:**       | Copyright (c) 2015-2016, Xabier de Zuazo
| **Copyright:**       | Copyright (c) 2015, Onddo Labs, SL.
| **License:**         | Apache License, Version 2.0

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
# CHANGELOG for ssh_authorized_keys

All notable changes to the [`ssh_authorized_keys`](https://supermarket.chef.io/cookbooks/ssh_authorized_keys) Chef cookbook will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
- ...

## [1.0.0] - 2020-08-09
[![Build Status](https://img.shields.io/travis/zuazo/ssh_authorized_keys-cookbook/1.0.0.svg?style=flat)](https://travis-ci.org/zuazo/ssh_authorized_keys-cookbook)

### Added
- Chef 16 support (issues [#15](https://github.com/zuazo/ssh_authorized_keys-cookbook/pull/15) and [#16](https://github.com/zuazo/ssh_authorized_keys-cookbook/pull/16), thanks [Benjamin M. Hughes](https://github.com/bmhughes) and [Corey Hemminger](https://github.com/Stromweld)).
- Included Oracle support.

### Removed
- Drop Ruby `< 2.5` support.
- Drop Chef `< 14` support.

### Documentation Changes
- README: Fix broken badges

## [0.4.0] - 2017-06-12
[![Build Status](https://img.shields.io/travis/zuazo/ssh_authorized_keys-cookbook/0.4.0.svg?style=flat)](https://travis-ci.org/zuazo/ssh_authorized_keys-cookbook)

### Added
- Adding ability to disable assert ([issue #6](https://github.com/zuazo/ssh_authorized_keys-cookbook/pull/6), thanks [Anthony Caiafa](https://github.com/acaiafa)).
- Allowing direct key injection ([issue #7](https://github.com/zuazo/ssh_authorized_keys-cookbook/pull/7), thanks [Anthony Caiafa](https://github.com/acaiafa)).
- Add AIX as supported platform ([issue #8](https://github.com/zuazo/ssh_authorized_keys-cookbook/issues/8), thanks [Mike Veltman](https://github.com/MVNW)).
- metadata: Add `chef_version`.

### Removed
- Drop Ruby `< 2.3` support.
- Drop Chef `< 12` support.

### Documentation Changes
- README:
  - Add documentation badge.
  - Fix Parameters table.
- Add GitHub templates.
- CHANGELOG: Follow "Keep a CHANGELOG".

## [0.3.0] - 2015-08-14
[![Build Status](https://img.shields.io/travis/zuazo/ssh_authorized_keys-cookbook/0.3.0.svg?style=flat)](https://travis-ci.org/zuazo/ssh_authorized_keys-cookbook)

### Fixed
- Use *gid* from `Etc.getpwnam` (issue [#3](https://github.com/zuazo/ssh_authorized_keys-cookbook/pull/3), thanks [Ong Ming Yang](https://github.com/ongmingyang)).
- Recursively create *.ssh* directory if it does not exist (issue [#4](https://github.com/zuazo/ssh_authorized_keys-cookbook/pull/4), thanks [Ong Ming Yang](https://github.com/ongmingyang)).
- Update chef links to use *chef.io* domain.
- Update contact information and links after migration.

## [0.2.0] - 2015-05-27
[![Build Status](https://img.shields.io/travis/zuazo/ssh_authorized_keys-cookbook/0.2.0.svg?style=flat)](https://travis-ci.org/zuazo/ssh_authorized_keys-cookbook)

### Added
- Add openSUSE as supported platform.

### Fixed
- Sort keys always in the same order ([issue #2](https://github.com/zuazo/ssh_authorized_keys-cookbook/issues/2), thanks to [Chris Burroughs](https://github.com/cburroughs) for the idea).

## 0.1.0 - 2015-01-11
[![Build Status](https://img.shields.io/travis/zuazo/ssh_authorized_keys-cookbook/0.1.0.svg?style=flat)](https://travis-ci.org/zuazo/ssh_authorized_keys-cookbook)

- Initial release of `ssh_authorized_keys`.

[Unreleased]: https://github.com/zuazo/ssh_authorized_keys-cookbook/compare/1.0.0...HEAD
[1.0.0]: https://github.com/zuazo/ssh_authorized_keys-cookbook/compare/0.4.0...1.0.0
[0.4.0]: https://github.com/zuazo/ssh_authorized_keys-cookbook/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/zuazo/ssh_authorized_keys-cookbook/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/zuazo/ssh_authorized_keys-cookbook/compare/0.1.0...0.2.0


spec/spec_helper.rb
# encoding: UTF-8
#
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'libraries'))

require 'chefspec'
require 'chefspec/berkshelf'
require 'should_not/rspec'

require 'support/coverage'

RSpec.configure do |config|
  # Prohibit using the should syntax
  config.expect_with :rspec do |spec|
    spec.syntax = :expect
  end

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  # --seed 1234
  # config.order = 'random'
  # Library tests first (they are capitalized) to not interfere with coverage
  config.register_ordering(:global) do |list|
    list.sort_by(&:description)
  end

  # ChefSpec configuration
  config.log_level = :fatal
  config.color = true
  config.formatter = :documentation
  config.tty = true
  config.platform = 'ubuntu'
  config.version = '12.04'
end


spec/support/coverage.rb
# encoding: UTF-8
#
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if ENV['TRAVIS']
  require 'coveralls'
  Coveralls.wear!
else
  require 'simplecov'
  SimpleCov.start do
    add_group 'Libraries', '/libraries'
    add_group 'ChefSpec' do |src|
      %r{/spec/(recipes|resources|providers)}.match(src.filename)
    end
    add_group 'RSpec' do |src|
      %r{/spec/(unit|functional|integration|libraries)}.match(src.filename)
    end
    add_group 'RSpec Support', '/spec/support'
  end
end


spec/recipes/default_spec.rb
# encoding: UTF-8
#
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'spec_helper'

describe 'ssh_authorized_keys_test::default', order: :random do
  let(:chef_runner) { ChefSpec::SoloRunner.new }
  let(:chef_run) { chef_runner.converge(described_recipe) }

  it 'creates bob user' do
    expect(chef_run).to create_user('bob')
      .with_supports(manage_home: true)
      .with_home('/home/bob')
  end

  it 'creates bob group' do
    expect(chef_run).to create_group('bob')
      .with_members(%w(bob))
  end

  it 'creates bob2 user' do
    expect(chef_run).to create_user('bob2')
      .with_supports(manage_home: true)
      .with_home('/home/bob2')
  end

  it 'creates bob2 group' do
    expect(chef_run).to create_group('bob2')
      .with_members(%w(bob2))
  end

  it 'creates alice user' do
    expect(chef_run).to create_user('alice')
      .with_supports(manage_home: true)
      .with_home('/home/alice')
  end

  it 'creates alice group' do
    expect(chef_run).to create_group('alice')
      .with_members(%w(alice))
  end

  it 'creates ~bob/.ssh directory recursively' do
    expect(chef_run).to create_directory('/home/bob/.ssh')
      .with_recursive(true)
      .with_owner('bob')
      .with_group('bob')
      .with_mode('00700')
  end

  it 'creates ~bob/.ssh/authorized_keys file' do
    expect(chef_run).to create_template('/home/bob/.ssh/authorized_keys')
      .with_cookbook('ssh_authorized_keys')
      .with_source('authorized_keys.erb')
      .with_owner('bob')
      .with_group('bob')
      .with_mode('00600')
  end

  it 'authorizes bob@acme.com to login as bob' do
    expect(chef_run).to render_file('/home/bob/.ssh/authorized_keys')
      .with_content(%r{^ssh-rsa [A-Za-z0-9+/=]+ bob@acme\.com$})
  end

  it 'authorizes bob@home.com to login as bob' do
    expect(chef_run).to render_file('/home/bob/.ssh/authorized_keys')
      .with_content(%r{^ssh-rsa [A-Za-z0-9+/=]+ bob@home\.com comment$})
  end

  it 'creates ~bob2/.ssh directory recursively' do
    expect(chef_run).to create_directory('/home/bob2/.ssh')
      .with_recursive(true)
      .with_owner('bob2')
      .with_group('bob2')
      .with_mode('00700')
  end

  it 'creates ~bob2/.ssh/authorized_keys file' do
    expect(chef_run).to create_template('/home/bob2/.ssh/authorized_keys')
      .with_cookbook('ssh_authorized_keys')
      .with_source('authorized_keys.erb')
      .with_owner('bob2')
      .with_group('bob2')
      .with_mode('00600')
  end

  it 'authorizes bob@acme.com to login as bob2' do
    expect(chef_run).to render_file('/home/bob2/.ssh/authorized_keys')
      .with_content(%r{^ssh-rsa [A-Za-z0-9+/=]+ bob@acme\.com$})
  end

  it 'authorizes bob@home.com to login as bob2' do
    expect(chef_run).to render_file('/home/bob2/.ssh/authorized_keys')
      .with_content(%r{^ssh-rsa [A-Za-z0-9+/=]+ bob@home\.com comment$})
  end

  it 'writes bob2 keys in the same order as bob keys' do
    file1 = chef_run.template('/home/bob/.ssh/authorized_keys')
    content1 = ChefSpec::Renderer.new(chef_run, file1).content
    expect(chef_run).to render_file('/home/bob2/.ssh/authorized_keys')
      .with_content(content1)
  end

  it 'creates ~alice/.ssh directory recursively' do
    expect(chef_run).to create_directory('/home/alice/.ssh')
      .with_recursive(true)
      .with_owner('alice')
      .with_group('alice')
      .with_mode('00700')
  end

  it 'creates ~alice/.ssh/authorized_keys file' do
    expect(chef_run).to create_template('/home/alice/.ssh/authorized_keys')
      .with_cookbook('ssh_authorized_keys')
      .with_source('authorized_keys.erb')
      .with_owner('alice')
      .with_group('alice')
      .with_mode('00600')
  end

  it 'authorizes alice@acme.com to login as alice including options' do
    options =
      'no-port-forwarding,no-agent-forwarding,no-X11-forwarding,'\
      'command="echo \'Please login as the user \"bob\" rather than the user '\
      '\"alice\".\';echo;sleep 10"'
    expect(chef_run).to render_file('/home/alice/.ssh/authorized_keys')
      .with_content(
        %r{^#{Regexp.escape(options)} ssh-rsa [A-Za-z0-9+/=]+ alice@acme\.com$}
      )
  end

  it 'creates ~root/.ssh directory' do
    expect(chef_run).to create_directory('/root/.ssh')
      .with_mode('00700')
  end

  it 'creates ~root/.ssh/key2 file' do
    expect(chef_run).to create_file('/root/.ssh/key2')
      .with_mode('00600')
  end
end


metadata.rb
# encoding: UTF-8
#
# Cookbook:: ssh_authorized_keys
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015-2016 Xabier de Zuazo
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

name 'ssh_authorized_keys'
maintainer 'Xabier de Zuazo'
maintainer_email 'xabier@zuazo.org'
license 'Apache-2.0'
description 'Creates SSH authorized keys files in user home directories.'
# long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version '1.1.0' # WiP

source_url "https://github.com/zuazo/#{name}-cookbook"
issues_url "https://github.com/zuazo/#{name}-cookbook/issues"

chef_version '>= 14'

supports 'aix'
supports 'amazon'
supports 'debian'
supports 'centos'
supports 'fedora'
supports 'freebsd'
supports 'opensuse'
supports 'oracle'
supports 'redhat'
supports 'suse'
supports 'ubuntu'


resources/key.rb
# encoding: UTF-8
#
# Cookbook:: ssh_authorized_keys
# Resource:: key
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

resource_name :ssh_authorize_key
default_action :create

property :instance_name, String, name_property: true
property :user, [Integer, String]
property :group, [Integer, String]
property :home, String
property :key, String
property :keytype, String, default: 'ssh-rsa'
property :comment, String
property :options, Hash
property :validate_key, [true, false], default: true

include SshAuthorizedKeysCookbook::TemplateHelpers
include SshAuthorizedKeysCookbook::ResourceHelpers

action :create do
  new_resource.comment ||= new_resource.instance_name
  new_resource.group ||= user_group(new_resource.user)
  new_resource.home ||= user_home(new_resource.user)
  assert_user(new_resource.user)

  ssh_key = {
    options: new_resource.options,
    key: new_resource.key,
    validate_key: new_resource.validate_key,
    keytype: new_resource.keytype,
    comment: new_resource.comment,
  }

  assert_key(ssh_key[:key]) if new_resource.validate_key
  assert_keytype(ssh_key[:keytype])
  assert_comment(ssh_key[:comment])

  path = ::File.join(new_resource.home, '.ssh', 'authorized_keys')

  # Avoid CHEF-3694 warning for .ssh directory.
  dir = ::File.dirname(path)

  directory dir do
    recursive true
    owner new_resource.user
    group new_resource.group
    mode '00700'
  end

  with_run_context :root do
    edit_resource(:template, path) do
      cookbook 'ssh_authorized_keys'
      source 'authorized_keys.erb'
      owner new_resource.user
      group new_resource.group
      mode '00600'
      variables['keys'] ||= {}
      variables['keys'][ssh_key[:comment]] = ssh_key
      variables['keys'] = variables['keys'].sort.to_h
      action :nothing
      delayed_action :create
    end
  end
end


attributes/default.rb
# encoding: UTF-8
#
# Cookbook:: ssh_authorized_keys
# Attributes:: default
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['ssh_authorized_keys']['keytypes'] = %w(
  ecdsa-sha2-nistp256
  ecdsa-sha2-nistp384
  ecdsa-sha2-nistp521
  ssh-ed25519
  ssh-dss
  ssh-rsa
)


libraries/resource_helpers.rb
# encoding: UTF-8
#
# Cookbook:: ssh_authorized_keys
# Library:: resource_helpers
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015-2016 Xabier de Zuazo
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'etc'

# `ssh_authorized_keys` internal classes.
module SshAuthorizedKeysCookbook
  # Some helpers to use from `ssh_authorized_keys` cookbook resources and
  # definitions.
  #
  # @example
  #   self.class.send(:include, ::SshAuthorizedKeysCookbook::ResourceHelpers)
  #   user_home('vagrant') #=> "/home/vagrant"
  module ResourceHelpers
    unless defined?(::SshAuthorizedKeysCookbook::ResourceHelpers::SSH_KEY_REGEX)
      # Regular expression for SSH public keys in base64.
      SSH_KEY_REGEX = %r{
        ^(?:[A-Za-z0-9+\/]{4})*(?:
          [A-Za-z0-9+\/]{2}==
          |[A-Za-z0-9+\/]{3}=
          |[A-Za-z0-9+\/]{4}
        )$
      }x.freeze
    end

    # Asserts that the user name is correct.
    #
    # @param user [String] user name.
    # @raise [Chef::Exceptions::ValidationFailed] if the user name is wrong.
    # @return void
    def assert_user(user)
      return if user.is_a?(String) && !user.empty?
      raise Chef::Exceptions::ValidationFailed,
            'ssh_authorize_key: user parameter must be a valid system user! '\
            "You passed #{user.inspect}."
    end

    # Asserts that the SSH public key is correct.
    #
    # @param key [String] public key in base64.
    # @raise [Chef::Exceptions::ValidationFailed] if the key is wrong.
    # @return void
    def assert_key(key)
      return if key.is_a?(String) && !SSH_KEY_REGEX.match(key).nil?
      raise Chef::Exceptions::ValidationFailed,
            'ssh_authorize_key: key parameter must be a valid SSH public key! '\
            "You passed #{key.inspect}."
    end

    # Returns allowed SSH key types list.
    #
    # @return [Array<String>] key types list.
    # @example
    #   allowed_keytypes
    #   #=> ["ecdsa-sha2-nistp256", "ecdsa-sha2-nistp384",
    #   #    "ecdsa-sha2-nistp521", "ssh-ed25519", "ssh-dss", "ssh-rsa"]
    def allowed_keytypes
      node['ssh_authorized_keys']['keytypes']
    end

    # Asserts that the SSH key type is correct.
    #
    # @param keytype [String] key type. Supported types are `'ssh-rsa'`,
    #   `'ssh-dss'`, `'ssh-ed25519'`, `'ecdsa-sha2-nistp521'`,
    #   `'ecdsa-sha2-nistp384'` and `'ecdsa-sha2-nistp256'`.
    # @raise [Chef::Exceptions::ValidationFailed] if the keytype is wrong.
    # @return void
    def assert_keytype(keytype)
      return if allowed_keytypes.include?(keytype)
      raise Chef::Exceptions::ValidationFailed,
            'ssh_authorize_key: keytype parameter must be equal to one of: '\
            "#{allowed_keytypes.join(', ')}! You passed #{keytype.inspect}."
    end

    # Asserts that the key comment is correct.
    #
    # @param comment [String] key comment or description.
    # @raise [Chef::Exceptions::ValidationFailed] if the comment is wrong.
    # @return void
    def assert_comment(comment)
      if comment.is_a?(String) && !comment.empty? && !comment.include?("\n")
        return
      end
      raise Chef::Exceptions::ValidationFailed,
            'ssh_authorize_key: comment parameter must be valid! You passed '\
            "#{comment.inspect}."
    end

    # Returns the home directory of a system user.
    #
    # If the user does not exist, it returns `"/home/#{user}"` as the home
    # directory and emits a Chef warning.
    #
    # @param user [String] user name.
    # @return [String] home directory.
    # @example
    #   user_home('root') #=> "/root"
    #   user_home('mail') #=> "/var/mail"
    #   user_home('bob')
    #   #WARN: ssh_authorize_key: User bob not found at compile time, perhaps
    #   #you should specify a home path. I will use "/home/bob" for now.
    #   #=> "/home/bob"
    def user_home(user)
      Etc.getpwnam(user).dir
    rescue ArgumentError
      home = ::File.join('', 'home', user)
      Chef::Log.warn(
        "ssh_authorize_key: User #{user} not found at compile time, perhaps "\
        "you should specify a home path. I will use #{home.inspect} for now."
      )
      home
    end

    # Returns the group of a system user.
    #
    # @param user [String] user name.
    # @return [Integer] gid.
    # @example
    #   user_group('root') #=> 0
    def user_group(user)
      Etc.getpwnam(user).gid
    rescue ArgumentError
      Chef::Log.warn(
        "ssh_authorize_key: User #{user} not found at compile time, perhaps "\
        "you should specify a default group. I will use #{user} for now."
      )
      user
    end
  end
end


resources/ssh_authorized_key.rb
# encoding: UTF-8
#
# Cookbook:: ssh_authorized_keys
# Resource:: ssh_authorize_key
# Author:: Ben Hughes
# Author:: Corey Hemminger
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

provides :ssh_authorize_key
resource_name :ssh_authorize_key
default_action :create

description 'Manages ssh keys in users authorized_keys file'

property :instance_name, String,
         name_property: true,
         description: 'Resource name'

property :user, [Integer, String],
         required: true,
         description: 'System user **(required)**'

property :group, [Integer, String],
         description: 'System group'

property :home, String,
         description: 'System user home path'

property :key, String,
         required: true,
         description: 'SSH public key in base64 **(required)**'

property :keytype, String,
         default: 'ssh-rsa',
         description: 'SSH key type'

property :comment, String,
         name_property: true,
         description: 'SSH key comment'

property :options, Hash,
         description: 'SSH key options as a hash'

property :validate_key, [true, false],
         default: true,
         description: 'Enable/Disable assert_key'

action :create do
  assert_user(new_resource.user)
  new_resource.comment ||= new_resource.instance_name
  new_resource.group ||= user_group(new_resource.user)
  new_resource.home ||= user_home(new_resource.user)

  ssh_key = {
    options: new_resource.options,
    key: new_resource.key,
    validate_key: new_resource.validate_key,
    keytype: new_resource.keytype,
    comment: new_resource.comment,
  }

  assert_key(ssh_key[:key]) if new_resource.validate_key
  assert_keytype(ssh_key[:keytype])
  assert_comment(ssh_key[:comment])

  path = ::File.join(new_resource.home, '.ssh', 'authorized_keys')

  # Avoid CHEF-3694 warning for .ssh directory.
  dir = ::File.dirname(path)

  directory dir do
    recursive true
    owner new_resource.user
    group new_resource.group
    mode '00700'
  end

  # Accumulator Pattern:
  # https://blog.dnsimple.com/2017/10/chef-accumulators/
  with_run_context :root do
    edit_resource(:template, path) do
      cookbook 'ssh_authorized_keys'
      source 'authorized_keys.erb'
      owner new_resource.user
      group new_resource.group
      mode '00600'
      variables['keys'] ||= {}
      variables['keys'][ssh_key[:comment]] = ssh_key
      variables['keys'] = variables['keys'].sort.to_h
      action :nothing
      delayed_action :create
    end
  end
end

action :remove do
  description 'Remove ssh keys from users authorized_keys file'
  # TODO: Create remove logic
end

action_class do
  include SshAuthorizedKeysCookbook::TemplateHelpers
  include SshAuthorizedKeysCookbook::ResourceHelpers
end
